### PR TITLE
Fix Iframe #postMessage Error

### PR DESCRIPTION
**Pull request by jridgewell**
Fixes https://github.com/ampproject/amphtml/issues/4034.

Never post to an iframe that is no longer active. We had three cases:

1. When the iframe requests a resize, but is removed before the resize completes. In that case, we'd cue up a promise to eventually send a `embed-size-X` message but send it to a non-existent iframe window.
2. When a ad's sub-iframes subscribe to the `embed-state` message, but the sub-iframe (or parent) is later removed.
3. When an ad's or iframe's sub-iframes subscribe to the `intersection` message, but the sub-iframe is removed.

### CODE REVIEW

File: src/utils/array.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * A bit like Array#filter, but removes elements that filter true from the
+ * array. Returns the filtered items.
+ *
+ * @param {!Array<T>} array
+ * @param {function(T, number, !Array<T>):boolean} filter
+ * @return {!Array<T>}
+ * @template T
+ */
+export function filterSplice(array, filter) {
```
Comment (dvoytenko) [inline src/utils/array.js]: This needs tests

**Comment by dvoytenko** [comment]
LGTM, but tests are needed for array utils

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 2]
```diff
@@ -169,16 +170,16 @@ export class AmpAdApiHandler {
...
   /** @override  */
   unlayoutCallback() {
-    this.unlisteners_.forEach(unlistener => unlistener());
-    this.unlisteners_ = [];
     if (this.iframe_) {
+      this.unlisteners_.forEach(unlistener => unlistener());
+      this.unlisteners_.length = 0;
+      this.embedStateApi_.destroy();
+      this.embedStateApi_ = null;
+      this.intersectionObserver_.destroy();
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: `this.iframe_` could be set to `null` by `noContentHandler`. We need to unlisten and destroy `intersectionObserver_` for that as well.

File: src/utils/array.js (hunk 1 above)
Comment (jridgewell) [inline src/utils/array.js]: Added.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 2 above)
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: That's a different `this`, but we should be destroying when there's no content. I'll add that.

**Pull request merged**